File: apps/api/pkg/domain/route.go
package domain

import (
	"sort"
	"strings"
	"time"
)

// The route a trip is built on (DEV_SPEC M1 — A1.3).
//
// This file is the server-side twin of `apps/web/lib/data/route.ts`; the two
// must agree, because the entry flow computes the route in the browser before
// the trip exists and the trip room reads it back from here afterwards.
//
// The whole point is that a route answers questions a list of city names never
// could. "โซล, อูเอโนะ" left three of them open — is that two countries, how
// many nights in each, and how do you get between them. Legs answer all three:
// each arrival opens a stop, the next departure closes it, and the nights in
// between are the plan.

// Leg is one hop of the route, already validated by the caller.
type Leg struct {
	Direction  string // out | inter | back
	Mode       string // flight | ground
	FlightNo   string
	DepAirport string // IATA
	ArrAirport string // IATA
	DepDate    string // "2026-12-04", required
	DepTime    string // "18:00", may be empty
	ArrDate    string // defaults to DepDate when empty
	ArrTime    string // "08:05", may be empty
}

// Airport is the slice of the index the route needs. It is an interface-free
// struct so pkg/domain keeps depending on nothing.
type Airport struct {
	IATA        string `json:"iata"`
	City        string `json:"city"`
	CityTH      string `json:"city_th"`
	CountryCode string `json:"country_code"`
	Country     string `json:"country"`
	CountryTH   string `json:"country_th"`
}

// Stop is one place the group actually stays, with the nights it gets.
type Stop struct {
	Airport     string `json:"airport"`
	City        string `json:"city"`
	CountryCode string `json:"country_code"`
	Country     string `json:"country"`
	ArriveDate  string `json:"arrive_date"`
	ArriveTime  string `json:"arrive_time"`
	DepartDate  string `json:"depart_date"`
	DepartTime  string `json:"depart_time"`
	Nights      int    `json:"nights"`
	// True while no leg leaves this stop — a one-way route, or a route still
	// being filled in.
	Open bool `json:"open"`
}

// CountryStay aggregates the stops that share a country, in visit order.
type CountryStay struct {
	Code   string `json:"code"`
	Name   string `json:"name"`
	Cities string `json:"cities"`
	Nights int    `json:"nights"`
}

// Route is everything derived from the legs.
type Route struct {
	HomeAirport string        `json:"home_airport"`
	Stops       []Stop        `json:"stops"`
	Countries   []CountryStay `json:"countries"`
	StartDate   string        `json:"start_date"`
	EndDate     string        `json:"end_date"`
	Days        int           `json:"days"`
	Nights      int           `json:"nights"`
	RoundTrip   bool          `json:"round_trip"`
}

// BuildRoute derives the frame from the legs. `lookup` resolves a IATA code;
// an unknown code still produces a stop, named after the code itself, because
// dropping a leg would silently shorten the trip.
func BuildRoute(legs []Leg, lookup func(iata string) *Airport) Route {
	route := Route{Stops: []Stop{}, Countries: []CountryStay{}}
	legs = sortedLegs(legs)
	if len(legs) == 0 {
		return route
	}

	route.HomeAirport = strings.ToUpper(legs[0].DepAirport)
	route.StartDate = legs[0].DepDate
	route.EndDate = arrivalDate(legs[len(legs)-1])

	for i, leg := range legs {
		arr := strings.ToUpper(leg.ArrAirport)

		// The last leg landing back where the trip started is the way home, not
		// a stop. Anything else is somewhere the group wakes up.
		if i == len(legs)-1 && arr == route.HomeAirport {
			route.RoundTrip = true
			continue
		}

		stop := Stop{
			Airport:    arr,
			ArriveDate: arrivalDate(leg),
			ArriveTime: leg.ArrTime,
			Open:       true,
		}
		if a := lookup(arr); a != nil {
			stop.City = firstNonEmpty(a.CityTH, a.City, arr)
			stop.CountryCode = a.CountryCode
			stop.Country = firstNonEmpty(a.CountryTH, a.Country, a.CountryCode)
		} else {
			stop.City = arr
		}

		if i+1 < len(legs) {
			next := legs[i+1]
			stop.DepartDate = next.DepDate
			stop.DepartTime = next.DepTime
			stop.Nights = nightsBetween(stop.ArriveDate, stop.DepartDate)
			stop.Open = false
		}

		route.Stops = append(route.Stops, stop)
	}

	route.Days = nightsBetween(route.StartDate, route.EndDate) + 1
	route.Nights = maxOf(route.Days-1, 0)
	route.Countries = countryStays(route.Stops)
	return route
}

// Cities lists the destinations in visit order, without repeats — what the trip
// frame stores in destination_cities.
func (r Route) Cities() []string {
	seen := map[string]bool{}
	out := make([]string, 0, len(r.Stops))
	for _, s := range r.Stops {
		if s.City == "" || seen[s.City] {
			continue
		}
		seen[s.City] = true
		out = append(out, s.City)
	}
	return out
}

// PrimaryCountry is the country with the most nights — the one the trip is
// priced and named after when a route crosses several.
func (r Route) PrimaryCountry() string {
	best := ""
	nights := -1
	for _, c := range r.Countries {
		if c.Nights > nights {
			best, nights = c.Code, c.Nights
		}
	}
	return best
}

/* ------------------------------------------------------------------ utils -- */

func sortedLegs(legs []Leg) []Leg {
	out := make([]Leg, 0, len(legs))
	for _, l := range legs {
		if strings.TrimSpace(l.DepAirport) == "" || strings.TrimSpace(l.ArrAirport) == "" {
			continue
		}
		if strings.TrimSpace(l.DepDate) == "" {
			continue
		}
		out = append(out, l)
	}
	sort.SliceStable(out, func(a, b int) bool {
		if out[a].DepDate != out[b].DepDate {
			return out[a].DepDate < out[b].DepDate
		}
		return out[a].DepTime < out[b].DepTime
	})
	return out
}

func arrivalDate(l Leg) string {
	if strings.TrimSpace(l.ArrDate) != "" {
		return l.ArrDate
	}
	return l.DepDate
}

func countryStays(stops []Stop) []CountryStay {
	order := make([]string, 0, len(stops))
	byCode := map[string]*CountryStay{}

	for _, s := range stops {
		code := s.CountryCode
		if code == "" {
			code = s.Airport
		}
		stay, ok := byCode[code]
		if !ok {
			stay = &CountryStay{Code: code, Name: firstNonEmpty(s.Country, code)}
			byCode[code] = stay
			order = append(order, code)
		}
		stay.Nights += s.Nights
		if stay.Cities == "" {
			stay.Cities = s.City
		} else if !strings.Contains(stay.Cities, s.City) {
			stay.Cities += " · " + s.City
		}
	}

	out := make([]CountryStay, 0, len(order))
	for _, code := range order {
		out = append(out, *byCode[code])
	}
	return out
}

func nightsBetween(from, to string) int {
	start, err1 := time.Parse(isoDate, from)
	end, err2 := time.Parse(isoDate, to)
	if err1 != nil || err2 != nil {
		return 0
	}
	return maxOf(int(end.Sub(start).Hours()/24), 0)
}

func firstNonEmpty(values ...string) string {
	for _, v := range values {
		if strings.TrimSpace(v) != "" {
			return v
		}
	}
	return ""
}

func maxOf(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: apps/web/lib/data/route.ts
import { daysBetween, thaiDate } from './domain';
import type { Airport, CountryStay, FlightLeg, FlightLegInput, RouteStop, TripRoute } from './types';

/**
 * The route a trip is built on (M1 — A1.3).
 *
 * Twin of `apps/api/pkg/domain/route.go`; the two must agree, because the entry
 * flow builds the route in the browser before the trip exists and the trip room
 * reads it back from the API afterwards.
 *
 * The reason this exists at all: a list of city names could not answer the
 * three questions that decide a plan. "โซล, อูเอโนะ" left all three open — is
 * that two countries, how many nights in each, and how do you get between them.
 * Legs answer all three. Each arrival opens a stop, the next departure closes
 * it, and the nights in between are what the planner has to fill.
 */

type Lookup = (iata: string) => Airport | null | undefined;

/** Accepts saved legs and legs still being typed. */
type AnyLeg = FlightLeg | FlightLegInput;

export function buildRoute(legs: AnyLeg[], lookup: Lookup): TripRoute {
  const ordered = sortLegs(legs);
  const empty: TripRoute = {
    flights: [],
    stops: [],
    countries: [],
    homeAirport: '',
    startDate: '',
    endDate: '',
    days: 0,
    nights: 0,
    roundTrip: false,
  };
  if (ordered.length === 0) return empty;

  const homeAirport = ordered[0]!.from.toUpperCase();
  const startDate = ordered[0]!.depDate;
  const endDate = arrivalDate(ordered[ordered.length - 1]!);

  const stops: RouteStop[] = [];
  let roundTrip = false;

  ordered.forEach((leg, i) => {
    const arr = leg.to.toUpperCase();

    // The last leg landing back where the trip started is the way home, not a
    // place anyone wakes up.
    if (i === ordered.length - 1 && arr === homeAirport) {
      roundTrip = true;
      return;
    }

    const airport = lookup(arr);
    const next = ordered[i + 1];

    stops.push({
      airport: arr,
      city: airport ? (airport.cityTh || airport.city) : arr,
      countryCode: airport?.countryCode ?? '',
      country: airport ? (airport.countryTh || airport.country) : '',
      arriveDate: arrivalDate(leg),
      arriveTime: leg.arrTime,
      departDate: next?.depDate,
      departTime: next?.depTime,
      nights: next ? nightsBetween(arrivalDate(leg), next.depDate) : 0,
      open: !next,
    });
  });

  const days = nightsBetween(startDate, endDate) + 1;

  return {
    flights: ordered.filter(isSaved),
    stops,
    countries: countryStays(stops),
    homeAirport,
    startDate,
    endDate,
    days,
    nights: Math.max(days - 1, 0),
    roundTrip,
  };
}

/** Destinations in visit order, without repeats — what the trip frame stores. */
export function routeCities(route: TripRoute) {
  return [...new Set(route.stops.map((s) => s.city).filter(Boolean))];
}

/* --------------------------------------------------------------- warnings -- */

export type RouteWarningLevel = 'info' | 'warn';

export interface RouteWarning {
  id: string;
  level: RouteWarningLevel;
  text: string;
}

/**
 * What the route builder says out loud while someone types.
 *
 * These are the answers to the questions the old city picker could not answer,
 * so they are deliberately concrete: which two countries, how many nights, and
 * which gap has no leg covering it.
 */
export function routeWarnings(route: TripRoute, legs: AnyLeg[]): RouteWarning[] {
  const out: RouteWarning[] = [];
  const ordered = sortLegs(legs);
  if (ordered.length === 0) return out;

  // A hole in the route: the group is somewhere the next leg does not leave
  // from, and nothing in the trip says how they got across.
  for (let i = 0; i + 1 < ordered.length; i += 1) {
    const landed = ordered[i]!.to.toUpperCase();
    const leaves = ordered[i + 1]!.from.toUpperCase();
    if (landed !== leaves) {
      out.push({
        id: `gap-${landed}-${leaves}`,
        level: 'warn',
        text: `ลง ${landed} แต่เที่ยวถัดไปออกจาก ${leaves} — เพิ่มขาระหว่างเมือง (บินหรือไปเอง) ให้ครบก่อน แพลนจะได้ไม่ขาดช่วง`,
      });
    }
  }

  if (route.countries.length > 1) {
    const line = route.countries.map((c) => `${c.name} ${c.nights} คืน`).join(' · ');
    out.push({
      id: 'multi-country',
      level: 'info',
      text: `ทริปนี้ข้าม ${route.countries.length} ประเทศ: ${line} — ROVE จะแยกแพลนเป็นช่วงตามประเทศให้`,
    });
  }

  // A country reached by air and left the next morning. Legal, usually a
  // mistake, and always worth saying before the plan is drafted.
  for (const stop of route.stops) {
    if (!stop.open && stop.nights <= 1) {
      out.push({
        id: `short-${stop.airport}`,
        level: 'warn',
        text: `${stop.city} ได้แค่ ${stop.nights} คืน — เผื่อวันเพิ่มหรือตัดเมืองนี้ออกดีกว่าไหม`,
      });
    }
  }

  const last = route.stops[route.stops.length - 1];
  if (last?.open) {
    out.push({
      id: 'one-way',
      level: 'info',
      text: `ยังไม่มีขากลับจาก ${last.city} — ใส่ทีหลังได้ แพลนจะจบที่วันที่ใส่ล่าสุดไปก่อน`,
    });
  }

  return out;
}

/* ---------------------------------------------------------------- labels -- */

/** "โตเกียว 6 คืน · โซล 3 คืน" */
export function stayLabel(route: TripRoute) {
  return route.stops.map((s) => `${s.city} ${s.nights} คืน`).join(' · ');
}

/** "4 ธ.ค. ถึง 08:05" — how a leg reads on a card. */
export function legLabel(leg: AnyLeg) {
  const date = thaiDate(arrivalDate(leg));
  return leg.arrTime ? `${date} ถึง ${leg.arrTime} น.` : date;
}

/* ----------------------------------------------------------------- utils -- */

function isSaved(leg: AnyLeg): leg is FlightLeg {
  return typeof (leg as FlightLeg).id === 'string' && (leg as FlightLeg).id !== '';
}

/** Chronological, and without the half-typed legs the builder always has. */
function sortLegs(legs: AnyLeg[]): AnyLeg[] {
  return legs
    .filter((leg) => leg.from && leg.to && leg.depDate)
    .slice()
    .sort((a, b) => a.depDate.localeCompare(b.depDate) || (a.depTime ?? '').localeCompare(b.depTime ?? ''));
}

/** A red-eye is the exception: same-day arrival unless the leg says otherwise. */
function arrivalDate(leg: AnyLeg) {
  return leg.arrDate || leg.depDate;
}

/** daysBetween counts both ends; nights is the gap between them. */
function nightsBetween(from: string, to: string) {
  if (!from || !to) return 0;
  return Math.max(daysBetween(from, to) - 1, 0);
}

function countryStays(stops: RouteStop[]): CountryStay[] {
  const order: string[] = [];
  const byCode = new Map<string, CountryStay>();

  for (const stop of stops) {
    const code = stop.countryCode || stop.airport;
    let stay = byCode.get(code);
    if (!stay) {
      stay = { code, name: stop.country || code, cities: '', nights: 0 };
      byCode.set(code, stay);
      order.push(code);
    }
    stay.nights += stop.nights;
    if (!stay.cities) stay.cities = stop.city;
    else if (!stay.cities.includes(stop.city)) stay.cities += ` · ${stop.city}`;
  }

  return order.map((code) => byCode.get(code)!);
}
